src/utils/reqwest_resume/mod.rs

```diff
@@ -0,0 +1,295 @@
+    use hyper::service::{make_service_fn, service_fn};
+    use hyper::{Body, Request, Response, Server};
+
+    const BUFFER_LEN: usize = 4096 * 2;
+    const CHUNK_LEN: usize = 4096;
+
+    fn extract_range_start(value: &HeaderValue, total_len: usize) -> u64 {
+        let s = std::str::from_utf8(value.as_bytes()).unwrap();
+        let parse_ranges = parse_range_header(s).unwrap();
+        let range = parse_ranges.validate(total_len as u64).unwrap();
+        *range[0].start()
+    }
+
+    async fn hello(req: Request<Body>) -> Result<Response<Body>, Infallible> {
+        let buffer = [b'a'; BUFFER_LEN];
+
+        let body = if let Some(range) = req.headers().get(header::RANGE) {
+            let offset = extract_range_start(range, buffer.len());
+            let (mut sender, body) = Body::channel();
+
+            // Send the rest of the buffer
+            tokio::task::spawn(async move {
+                sender
+                    .send_data(Bytes::copy_from_slice(&buffer[offset as usize..]))
+                    .await
+                    .unwrap();
+            });
+            body
+        } else {
+            let (mut sender, body) = Body::channel();
+            tokio::task::spawn(async move {
+                sender
+                    .send_data(Bytes::copy_from_slice(&buffer[0..CHUNK_LEN]))
+                    .await
+                    .unwrap();
+                sleep(Duration::from_millis(100)).await;
+                // `abort` will close the connection with an error so we can test the
+                // resume functionality
+                sender.abort();
+            });
+            body
+        };
+
+        let mut response: Response<_> = Response::new(body);
+        response
+            .headers_mut()
+            .insert(header::ACCEPT_RANGES, HeaderValue::from_static("bytes"));
+        Ok(response)
+    }
+
+    async fn create_flaky_server(
+        port: u16,
+    ) -> tokio::task::JoinHandle<std::result::Result<(), hyper::Error>> {
+        // For every connection, we must make a `Service` to handle all
+        // incoming HTTP requests on said connection.
+
+        let make_svc = make_service_fn(|_conn| {
+            // This is the `Service` that will handle the connection.
+            // `service_fn` is a helper to convert a function that
+            // returns a Response into a `Service`.
+            async { Ok::<_, Infallible>(service_fn(hello)) }
+        });
+
+        let addr = ([127, 0, 0, 1], port).into();
+
+        let server = Server::bind(&addr).serve(make_svc);
+
+        println!("Listening on http://{}", addr);
+
+        tokio::task::spawn(server)
+    }
+
+    #[tokio::test]
+    pub async fn test_resumable_get() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+        let port = 3000;
+        create_flaky_server(port).await;
+
+        let resp = get(reqwest::Url::parse(&format!("http://localhost:{}", port)).unwrap()).await?;
```

Comment:
Will displaying the local addr of the server work here?

Reply:
I've removed the print.